File: LeetCode/1046.last-stone-weight.cpp
/*
 * @lc app=leetcode id=1046 lang=cpp
 *
 * [1046] Last Stone Weight
 */
#include <iostream>
#include <vector>
#include <queue>
#include <math.h>
using namespace std;

// @lc code=start

class Solution
{
public:
    int lastStoneWeight(vector<int> &stones)
    {
        priority_queue<int, vector<int>> mm;
        for (auto i : stones)
        {
            mm.push(i);
        }
        this->sol(mm);
        return mm.empty() ? 0 : mm.top();
    }

    void sol(priority_queue<int, vector<int>> &mm)
    {
        if (mm.size() < 2)
        {
            return;
        };

        int x = mm.top();
        mm.pop();
        int y = mm.top();
        mm.pop();
        int diff = std::abs(x - y);
        if (diff != 0)
        {
            mm.push(diff);
        }

        this->sol(mm);
    }
};

// @lc code=end

File: LeetCode/1046.last-stone-weight.py
#
# @lc app=leetcode id=1046 lang=python3
#
# [1046] Last Stone Weight
#

from typing import List
import heapq 


# @lc code=start
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        if not stones:
            return 0
        maxHeap= []
        for i in stones:
            heapq.heappush(maxHeap, -i)
        while len(maxHeap) >= 2:
            y = -heapq.heappop(maxHeap)
            x = -heapq.heappop(maxHeap)
            if y == x:
                continue
            else:
                heapq.heappush(maxHeap, -abs(y - x))
        return -maxHeap[0] if  stones else 0
        
# @lc code=end

s = Solution()
print(s.lastStoneWeight([2,7,4,1,8,1]))